doc/qbk/4.0.BNF_parsing.qbk

```diff
@@ -0,0 +1,32 @@
+[/
+    Copyright (c) 2019 Vinnie Falco (vinnie.falco@gmail.com)
+    Copyright (c) 2021 Alan de Freitas (alandefreitas@gmail.com)
+
+    Distributed under the Boost Software License, Version 1.0. (See accompanying
+    file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+    Official repository: https://github.com/CPPAlliance/url
+]
+
+[section:charset BNF Parsing]
+
+The file [@../../../../boost/url/bnf/parse.hpp `<boost/url/bnf/parse.hpp>`] contains the basic overloads of the function __bnf__parse__ for parsing BNF elements.
+
+[heading Design of BNF elements]
+
+A BNF element implements the rules for parsing a BNF grammar rule and stores parsing results. A BNF element:
+
+* Is default constructible
+* Holds the result, usually as a reference, of parsing
+* Has a nested value_type
+* Has `value_type& operator*() const noexcept`
```

Comment:
The nested type and `operator*` is only needed to work with `list` and `range`, this is a work in progress.